feat: populate `requested_spec`

baszalmstra opened the pull request:
Fixes #1595

This PR attempts to add support for writing (and updating) the `requested_spec` field in `PrefixRecord`s when installing an environment.

This adds a breaking change where the `requested_spec` is changed from an `Option<String>` to an `Vec<String>`. I made this change because multiple matchspecs can be added for the same package. We could potentially also merge matchspecs but I dont think thats completely trivial. 

Another think we have to think about is whether the order of the matchspecs in the `requested_spec` is important. In theory their order could influence the solve but Im not sure whether or not that is important. At the moment the implementation retains the order of the matchspecs as specified in the input specs.

wolfv commented on crates/rattler/src/install/installer/mod.rs:
```diff
@@ -680,3 +717,390 @@ async fn populate_cache(
+        // Write the updated record back to disk if needed
+        if let Some(new_record) = updated_record {
+            let conda_meta_path = prefix.path().join("conda-meta");
+            let pkg_meta_path = format!(
```
we should really make a function for this :)

wolfv commented on crates/rattler/src/install/installer/mod.rs:
```diff
@@ -680,3 +717,390 @@ async fn populate_cache(
+    spec_mapping: &HashMap<PackageName, Vec<String>>,
+    prefix: &Prefix,
+) -> Result<(), InstallerError> {
+    for record in existing_records {
```
Should we use `rayon` to parallelize this loop?

baszalmstra replied:
Changed to use rayon!

baszalmstra commented:
> I think I am mainly wondering: is `requested_spec` also a `list[str]` in Conda? Just want to make sure that we create "standards-compliant" JSON files here (and `requested_spec` sounds singular).

@jaimergp and I discussed this in https://github.com/conda/rattler/issues/1595#issuecomment-3176710758 and decided the use a Vec. Note that it still serializes as string too if there is only one entry for compatibility.

Im not sure though if conda will be able to parse the record if the `requested_spec` becomes a list. Ill have to try.

baszalmstra commented:
I changed the implementation, we now have two fields in `PrefixRecord`.

* `requested_spec` is just the original field. This field has been marked as deprecated.
* `requested_specs` is a new field of type `Vec<String>`.

I modified the `Installer` class to migrate the `requested_spec` to `requested_specs` and always update it if needed.
